README.md
# Chainpoint Node

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

## Frequently Asked Questions

Need help? Looking for the [FAQ](https://github.com/chainpoint/chainpoint-node/wiki/Frequently-Asked-Questions)?

## About

A Chainpoint Node allows anyone to run a server that accepts hashes, anchors them to public blockchains, creates and verifies proofs, and participates in the Chainpoint Network.

Nodes communicate with Chainpoint Core, and gain eligibility to earn TNT by providing services to the Chainpoint Network.

To be eligible to earn TNT a Node must:

* register with a unique Ethereum address
* maintain a minimum balance of 5000 TNT at that address
* provide public network services
* pass audits and health checks from Chainpoint Core

__*Chainpoint Nodes that don't meet all these requirements consistently won't be eligible to earn TNT through periodic rewards.*__

### What does a Node do?

Chainpoint Nodes aggregate incoming hashes into a Merkle tree every five seconds. The Merkle root is submitted to Chainpoint Core for anchoring to public blockchains.

Nodes allow clients to retrieve proofs for hashes that were previously submitted.

Nodes maintain a public mirror of the Calendar. This allows *any* Node to verify *any* proof.

Nodes expose a public HTTP API. There is some additional documentation, and examples of using the Node's public HTTP API with `curl` or other clients that can be found on the Wiki:

[Node HTTP API](https://github.com/chainpoint/chainpoint-node/wiki/Node-HTTP-API)

## About the Technology

### How Does It Work?

Chainpoint Node is run as a `docker-compose` application. `docker-compose` is a tool for running multiple Docker containers as
an orchestrated application suite.

Docker and `docker-compose` allow distribution of binary images that can run
anywhere that Docker can run. This ensures that everyone, regardless of
platform, is running the same code.

Docker makes it easy to distribute and upgrade the Chainpoint Node
software.

### Software Components

When started, `docker-compose` will install and run some system components in the Docker virtual machine.

* Chainpoint Node (a Node.js API server + RocksDB)
* NTP Time server (only if NTP is not already running natively)

These applications are started as a group and should not interfere with any other software systems running on your server. We do recommend running a Node on a server dedicated to that task.

Each Node instance you want to run will need to be configured with:

* A dedicated Ethereum address
* Public IP address (ports `22` and `80` open)
* Minimum TNT balance

### System Requirements

The software should be able to be run on any system that supports Docker and Docker Compose tools and meets the minimal hardware requirements.

#### Hardware

The optimal hardware requirements for running a Node are relatively modest.

Pro (High Volume):

- `>= 4GB RAM`
- `>= 2 CPU Core`
- `>= 2GB swapfile`
- `128GB+ SSD`
- `Public IPv4 address`

Recommended:

- `>= 2GB RAM`
- `>= 2 CPU Core`
- `>= 2GB swapfile`
- `64GB+ SSD`
- `Public IPv4 address`

Minimum:

- `>= 1GB RAM`
- `>= 1 CPU Core`
- `>= 2GB swapfile`
- `25GB+ SSD`
- `Public IPv4 address`

Running a Node on a server with less than 1GB of RAM is [known to cause issues](https://github.com/chainpoint/chainpoint-node/wiki/Frequently-Asked-Questions#operating-and-monitoring-a-node).

Nodes have relatively modest requirements for RAM and CPU. Nodes that receive sustained high volumes of hashes will write some temporary "proof state" data to SSD (approximately 350MB per million hashes). It is recommended to provision a minimum of 25GB of SSD storage. This temporary data will be automatically pruned over time.

RocksDB, used internally for data storage, is optimized for SSD disks. It is recommended to use SSD disks instead of traditional hard disks whenever possible.

If large volumes of hashes are sent to your server it's recommended that you scale-up the system by adding more RAM and SSD disk storage. Alternatively, you can scale-out horizontally by running more Nodes.

It is not supported to run multiple instances of the Node software on a single physical host.

#### Operating System

The software has been tested and is supported on the following operating systems:

* `Ubuntu 16.04 LTS`
* `macOS Version 10.12.6+`

It will likely run on other operating systems that support Docker and Docker Compose, but support is not currently provided for those.

#### Docker & Docker Compose

Nodes have been developed and tested on the following software versions.

* `Docker version 18.05.0-ce, build f150324`
* `docker-compose version 1.21.2, build a133471`

## Installation

This software is designed to be simple to install and run
on supported systems. Please follow the instructions below
to get started.

The following instructions should run on any public host running the `Ubuntu 16.04 LTS` operating system.

### Prerequisites

Before you start, you will need:

* An Ethereum address that you have the private keys for. Exchange provided accounts are generally not supported. You should be able to unlock your account to send Ether or TNT using MyEtherWallet for example.

* You must have the minimum balance of 5000 TNT to run a Node, and those TNT must be assigned to the Ethereum address you'll use to identify your Node. You can check your TNT balance (in Grains, divide by `100000000` (10^8) for TNT balance) using the Etherscan.io [contract reading tool for our ERC20 smart contract](https://etherscan.io/address/0x08f5a9235b08173b7569f83645d2c7fb55e8ccd8#readContract) (input your address in the `balanceOf` field and click `Query`).

If you run into any issues operating your Node, the first best place to look for help is the [Frequently Asked Questions](https://github.com/chainpoint/chainpoint-node/wiki/Frequently-Asked-Questions) page on our Wiki.

### Start a Server

Your first step is to start a server and gain SSH access
to your account on that server. This is beyond the scope of this document. You will need:

* `root` access, or a user with `sudo` privileges
* Ubuntu 16.04 LTS OS

You should always choose the simplest Ubuntu 16.04 image provided by your hosting provider. "Desktop" versions, or those that run an enabled version of the `ufw` firewall by default are not recommended as there are known compatibility issues with Docker and system firewalls. 

Log in to your server via SSH and continue to the next step, installing Docker and `docker-compose`.

### Install Docker and Docker Compose

To make this process easy we have created a small script, designed to be run on `Ubuntu 16.04 LTS`, that will install all runtime dependencies with a simple one-line command:

```sh
curl -sSL https://chainpoint-node.storage.googleapis.com/setup.sh | bash
```

Since this command runs a shell script as a privileged user on your system we recommend you [examine it carefully](https://github.com/chainpoint/chainpoint-node/blob/master/scripts/setup.sh) before you run it.

Simply copy/paste the command into your terminal, logged in as the root user, or another that has sudo privileges, and it will:

* Install Docker
* Install Docker Compose
* Configure a 2GB swap file that will survive reboots
* Adjust system config related to swap performance
* Grant the ability for your local user to run Docker commands without using `sudo`
* Download this repository to your home folder
* Create a default `.env` environment file, ready for you to edit

**Important**: You should close your terminal SSH session and login again after running it to make sure that the changes in the script are fully applied. You do not need to reboot your server.

#### Manual Install

If you prefer to do things manually there are [good instructions available](https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-on-ubuntu-16-04) for installing Docker on an Ubuntu server.

There are also [official docs for installing Docker](https://docs.docker.com/engine/installation/) on other systems.

For some systems you will need to separately install `docker-compose`.

### Sync your System Clock with NTP

Linux servers use a system service known as the Network Time Protocol (NTP) for synchronizing themselves to extremely accurate time sources. It is quite easy to setup a server so that it will stay within milliseconds, or even microseconds, of international standard clocks.

For Chainpoint Nodes we recommend using the public NTP time servers provided by Google. This will ensure that your servers are as close to Chainpoint Core server time as possible. Your Node will be continuously audited to ensure that it is within tolerances (currently max 5 seconds). Please take a moment to setup NTP to ensure you stay eligible for rewards over time.

* [Google Public NTP](https://developers.google.com/time/)
* [How to Configure NTP for Use in the NTP Pool Project on Ubuntu 16.04](https://www.digitalocean.com/community/tutorials/how-to-configure-ntp-for-use-in-the-ntp-pool-project-on-ubuntu-16-04)

If your system is not running its own NTP daemon (many do so by default), a privileged Docker container will attempt to start and continuously sync time for the Node.

### Configure Your Node

Configuration is as simple as editing a single configuration file and providing two variables. We provide a sample configuration file in this repository called `.env.sample`.

The installation script, if you use it, will copy that file
to `~/chainpoint-node/.env` for you, ready to edit.

If you installed everything manually you will want to:

```sh
cd ~/chainpoint-node
cp .env.sample .env

# use your favorite editor:
vi .env
```

There are only two values that you may need to edit (comments removed for clarity):

```sh
NODE_TNT_ADDRESS=
CHAINPOINT_NODE_PUBLIC_URI=
```

`NODE_TNT_ADDRESS` : should be set to your Ethereum address that contains TNT balance. It will start with `0x` and have an additional 40 hex characters (`0-9, a-f, A-F`). This is the unique identifier for your Node.

`CHAINPOINT_NODE_PUBLIC_URI` : should be a URI where your Node can be publicly discovered and utilized by others. This might look like `http://10.1.1.20`. Your Node will run on port `80` over `http`. If provided, this address will be periodically audited by Chainpoint Core to ensure compliance with the rules for a healthy Node. If you leave this config value blank, it will be assumed that your Node is not publicly available, and you will not be eligible to earn TNT rewards.

Once running you should be able to visit `http://YOURIP/config` from another host on the Internet and see a JSON response.

### Node Firewall

The network security of a Node is solely the Node operator's responsibility. We do not manage or modify the security settings or firewall of your Node server.

Most hosting providers block all incoming traffic (other than SSH on port 22) by default. You will need to expose port 80 (HTTP) to the world if you want your Node to be available for
clients to connect to. This is true whether you run a non-public host behind a firewall
or a public host. If your provider
does not implement a block by default policy you are encouraged to either choose another provider
or configure a firewall on your Node server to block access to all ports with the
exception of port `80 (HTTP)` and port `22 (SSH)`.

It is discouraged to run local firewall software, such as `ufw`, on your Node server unless you are an *expert* user and understand how Docker interacts with firewalls. `ufw` and other similar tools are known to interfere with Docker.

### Run Your Node

Now its time to start your Node!

After finishing the configuration in the `.env` file and saving it make sure you are in the `~/chainpoint-node` directory and run `make`. This will show you some Makefile commands that are available to you. Some of the more important ones are:

* `make up` : start all services
* `make down` : stop all services
* `make upgrade` : upgrade to newest release in git and restart
* `make restart` : restart only the Node software, git unchanged.
* `make logs` : show, and tail, the `docker-compose` logfiles
* `make ps` : show the status of the running processes

The simplest step at this point is to run `make up`. This
will automatically pull down the appropriate Docker images
(which may take a few minutes the first time, when there is an upgrade available, or on slower networks) and start them. `docker-compose` will also
try to ensure that the Node keeps running, even if it was to crash (which is unlikely, its very stable).

If you log out of your SSH session it will have no effect on your Node as long as you start it as described above.

### Monitor Your Node

There are two primary ways to monitor the health and operation of your Node. An authenticated web console, and via the command line.

#### Node Web UI

Once your Node is running, you can visit the URL `http://<YOUR-NODE-IP-ADDRESS>` in a browser. This UI is not public by default and you will be prompted to authenticate. The default password for your Node UI is the same as the `NODE_TNT_ADDRESS` that you set in your `.env` config file.

The Node UI will provide you with information related to the current activity of your Node, as well as statistics provided by Core about the health of your Node as measured during the twice hourly audit process and information about the count of the active Nodes in the queue. The 'Activity' page data is updated every few seconds, and the Node health information on the 'About' page is updated twice hourly. Health data is considered private to the Node and is pushed to the Node as a cryptographically signed data packet. There is no public API that similar information can be retrieved from.

If you prefer to set your own password for your Node UI there are instructions provided in the `.env.sample` file.

#### Command Line

In a terminal SSH session run `make ps` to see the `docker` services that are running.

Run `make logs` to continuously tail the logfiles for all `docker-compose` managed services.

When you start your Node you'll see in the logs that your Node will attempt to register itself with one of our Chainpoint Core clusters and will be provided with a secret key.

The Node will then go through a process of downloading, and cryptographically verifying the entire Chainpoint Calendar. Every block will have its signature checked and will be stored locally. This process may take some time on first run as our Calendar grows. After initial sync, all incremental changes will also be pulled down to every Node, verified and stored.

If there are any problems you see in the logs, or if something is not working as expected, please [file a bug](https://github.com/chainpoint/chainpoint-node/issues) and provide as much information about the issue as possible.

### Sending Hashes to Your Node with the CLI

Now you should be fully up and running! You might want to try out your new Node with the [Chainpoint Command Line Interface (CLI)](https://github.com/chainpoint/chainpoint-cli).

Normally the CLI will auto-discover a Node to send hashes to. Once you have it installed, you can configure it to always use your Node if you prefer.

You can either modify the Node address in the `~/.chainpoint/cli.config` to set it permanently, or you can override the Node address every time you use it like this:

```sh
chp submit --server http://127.0.0.1 <hash>
```

### Stopping Your Node

You can stop your Node at any time with `make down`. You can
verify that everything is stopped with `make ps`.

### Node Authentication Key Backup/Restore

*tl;dr* : On-server backups are now performed automatically (`v1.5.3+`). Export your auth keys with `make print-auth-keys` and store those backups elsewhere!!!

This command will print the `HMAC` keys to the console `STDOUT` and also give you a handy one-liner `echo` command to help with restoring the auth key on another host.

The following info may be useful if at some point in the future you need to backup/restore a Node or run it on a new host server.

### About Authentication Keys

Once your Node starts and registers itself with Chainpoint Core a secret key, sometimes referred to
as an `HMAC` or `Auth` key, will be generated and shared with your Node. This key sharing will only ever
take place once, at the moment when you first register your Node. The key will be stored in your Node's
local database. Every time that your Node starts it will use this key to authenticate itself to
Chainpoint Core and update your registered public IP address as needed. The Node's database can store
multiple auth keys and will choose the matching one based on the Ethereum address you've configured
in the `NODE_TNT_ADDRESS` environment variable in your Node's `.env` file.

*WARNING* : If this secret key is lost:

* you cannot recover it by any other means
* there is no way to "reset" the auth key associated with your Ethereum (TNT) address
* you will never be able to start another Node using the same Ethereum address.

To avoid loss you will want to back it up and store it somewhere safe in case of accidental deletion.

#### Backup

As of v1.5.3+ Nodes will automatically perform backups when the Node registers with Core and upon each startup. This will create a backup file for every auth key in your database in
the `keys/backups` sub-directory of your Node. The filename of the backup is composed of
the combination of your ETH address and the current time in the form: `<ETH_ADDRESS>-<TIMESTAMP>.key`.
The contents of the file will be a single line of text which is the HMAC Auth key associated
with that address.

#### Restore

Restoration of a backup copy of an auth key is easy. Every time a Node starts it looks at the
contents of the `~/chainpoint-node/keys` directory. If it finds a file where the name has one
of the following two patterns it will read the auth key out of the file and automatically import it into your Node.

* `<ETH_ADDRESS>.key`
* `<ETH_ADDRESS>-<TIMESTAMP>.key`

The contents of the `.key` file is simply the HMAC auth key for the address on a single line.

A `.key` file can be created using the backup procedure above, or you can manually create it.
If you manually create it, you don't need to include a timestamp in the filename.

__Restore Manually Created File__

Create a backup file (e.g. by copying the output of `make print-auth-keys` to a password manager). Restart your node with `make restart` after you do this to import the new auth key. The `make print-auth-keys` command will now also print an echo statement that can be used like this to help restore a key:

```sh
# an example of creating a key backup file manually
# run this in your `chainpoint-node` dir.
echo -n "my-secret-auth-hmac-key" > keys/0xMYETHADDRESS.key
```

__Restore Previously Backed Up File__

You'll find backup files in the `~/chainpoint-node/keys/backups` directory.

Copy the backup file where the filename matches the `NODE_TNT_ADDRESS` you have configured in your `.env` file from the `~/chainpoint-node/keys/backups` directory to the `~/chainpoint-node/keys` directory in order to make it ready to import when the Node next restarts.

Restart your node with `make restart` after you place a `.key` file in the `keys/` dir.

__Verify A Restoration__

When your Node restarts, if it found a `.key` file in the `~/chainpoint-node/keys` directory it will try to restore the auth key in that file to your local database. If it succeeds you will see a line similar to the following in your `make logs` output:

```sh
INFO : Registration : Auth key saved to local storage : 0x<MYETHADDRESS>-<TIMESTAMP>.key
```

## Frequently Asked Questions

Answers to many questions that have been raised by Node operators, and helpful tips, can be found in our [FAQ](https://github.com/chainpoint/chainpoint-node/wiki/Frequently-Asked-Questions#operating-and-monitoring-a-node)

Please refer to this document before filing any issues.

## License

[Apache License, Version 2.0](https://opensource.org/licenses/Apache-2.0)

```text
Copyright (C) 2017-2018 Tierion

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

## Thank You!

Thank you for being an active participant in the Chainpoint Network and for your interest in running a Chainpoint Node. We couldn't do it without you!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.5.4] - 2018-12-20

### Changes

- Upgraded Node.js base Docker image to Node.js `10.14.2` (LTS)
- Removed deprecated `make backup-auth-keys` Makefile target. Use `make print-auth-keys` instead.
- Removed all references to PostgreSQL, which is no longer used, from `docker-compose.yaml`.
- Removed all references to Redis, which is no longer used, from `docker-compose.yaml`.

### Added

- Optional experimental TLS/HTTPS support and self-signed certificate generation. Requires port 443 open for use. Port 80 will continue to operate the same as before.
- Retry mechanism on failure to submit an aggregated Merkle Root to Core.
- Added `block_id` & `block_value` to the `/verify` endpoint as a convenience when verifying a proof
- Community provided Korean translation of the `README.md` under the `community/translations/*` folder.


## [1.5.3] - 2018-08-20

### Changes

- Migrate and remove all PostgreSQL data to RocksDB. Redis and PostgreSQL containers will be removed in a future release.
- No longer start `chainpoint-ntpd` if `timesyncd` or `ntpd` is detected on Linux, or `timed`, or `pacemaker` on macOS. Previously only `ntpd` was tested for.
- Perform automatic backups of Auth (HMAC) keys to `keys/backups` directory on Node registration or restart.
- Deprecated `make backup-auth-keys` target as backups are now automatic.
- Added `echo` command to `make print-auth-keys` output. This provides easier to use one-line command to restore a backup key to the `keys` dir. Run this `echo` command to place a `.key` file in the keys directory with proper Auth Key (HMAC) content.
- Removed Redis and PostgreSQL related `make` targets as they are no longer needed.

### Added

- Send a `User-Agent` string with upstream requests.

### Fixed

- Updated NPM dependencies.


## [1.5.2] - 2018-07-30

### Changes

- Write ephemeral proof state to RocksDB (formerly Redis). This provides a significant increase in sustained inbound hash rate while consuming disk space instead of RAM. The ephemeral data written to RocksDB is auto-pruned after 24 hours. The Redis database will be removed from the Node in a future release.
- Upgrade to PostgreSQL 9.6.9
- Upgrade to Redis 4.0.10
- Upgrade Node.js (`node:8.9.0-alpine` -> `node:8.11.3-stretch`) Docker container base
- Reduce HTTP calls to `PUT /nodes` if request body matches previous successful `PUT`.
- Certain caching functions, which contain small amounts of volatile data, now store their data in RAM instead of Redis.

### Added

- Improved handling of processing time hints from Chainpoint Core.
- Auto-migrate old Redis proof state and other Redis data to RocksDB on initial startup.

### Fixed

- Node UI React application internal re-factoring and bugfixes.
- Ensure `validator.isIP()` receives a string avoiding rare crash in `1.5.1` release.
- Re-create `/keys/backups` directories if previously deleted.
- Improve detection, handling, and UI messaging in Brave browser when third-party cookies and local storage are disabled.
- Trim whitespace from the list of hashes provided to `POST /hashes` using the `hashids` request header.
- Further enforce Node operation on HTTP port 80.

## [1.5.1] - 2018-07-03

### Changes

- Replaced Node local firewall with custom updatable IP filter. This eliminates the need for the Node to schedule and perform automatic restarts.
- If a Node receives an HTTP `426 Upgrade Required` error from Core, indicating that the Node is running too old a version of the Node software to update its registration, it will no longer retry a request that cannot succeed. It will log an error message to the logs and exit.
- During registration, if an HTTP `409 Conflict` error is thrown as a result of the Ethereum address or public URI already being registered, the Node will no longer retry and will exit.
- When a Node is configured password with `CHAINPOINT_NODE_UI_PASSWORD=false` the Node UI will bypass the login screen and redirect the user directly to dashboard.

### Fixed

- An issue with persisting Redux data to browser local storage which sometimes resulted in the Node UI appearing 'stuck' and not displaying the most current data.
- Enabled authentication form submit via [enter] key, only validate after form submitted.

## [1.5.0] - 2018-06-27

### Changes

- Reduce retry count for failed registrations to 3.
- Nodes that fail to register will no longer exit after retries. The Node will continue to run, allowing access to the Node UI.
- Allow adjustment of Node hash aggregation period based on Core config.
- Update Chainpoint library dependencies `merkle-tools` and `chainpoint-parse` to current versions.

### Added

- New Makefile target for `make print-auth-keys` which will print the filename and auth key contents of each file in the `keys/backups` directory. This allows for easier copy/paste of backup key contents as an alternative to exploring the files in the `keys/backups` directory.
- Auth keys will be automatically backed up to the local drive on registration.
- Log some additional details about the cause(s) of a failed audit to the `make logs` output.

### Fixed

- Specifying `false` as a Boolean or String to `CHAINPOINT_NODE_UI_PASSWORD` in `.env` configuration will now work as expected.
- Update to Node UI to correctly display Node ms delta.
- Simplify and enhance performance of Core proof retrieval.

## [1.4.2] - 2018-05-08

### Changed

- `make clear-containers` will now only apply to Chainpoint related containers.
- `make clear-containers` no longer uses `sudo`. Assumes you have setup system so that logged-in user has permissions to docker commands without use of `sudo`.

scripts/upgrade.sh
#!/usr/bin/env bash
set -e

# Copyright 2018 Tierion
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#     http://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# You can run this upgrade script manually, by copying it to the host,
# or by issuing this curl command. Since this command pipes the script
# directly into a bash shell you should examine the script before running.
#
#   curl -sSL https://chainpoint-node.storage.googleapis.com/upgrade.sh | bash
#

# Don't run if seemingly outside of chainpoint-node dir.
if [ -f /docker-compose.yaml ]; then
    echo "This script should be run in the chainpoint-node dir. Exiting!"
    exit 0
fi

echo '#################################################'
echo 'Ensure Node is running on supported OS'
echo '#################################################'

if [ "$(. /etc/os-release; echo $NAME)" != "Ubuntu" ]; then
  echo "Looks like you are not running this on an Ubuntu OS. Exiting!"
  exit 1
fi

echo '#################################################'
echo 'Bring down existing Node and any orphans'
echo '#################################################'

docker-compose down --remove-orphans

echo '#################################################'
echo 'Clear old chainpoint-containers'
echo '#################################################'

containers=$(docker ps -aq -f "label=org.chainpoint.service");
if [ "${containers}" != "" ]; then
    echo "flushing docker containers...";
    docker stop ${containers};
    docker rm ${containers};
fi

echo '#################################################'
echo 'Upgrade docker-compose as needed'
echo '#################################################'

sudo mkdir -p /usr/local/bin
curl -sSL https://chainpoint-node.storage.googleapis.com/docker-compose-install.sh | bash

echo '#################################################'
echo 'Pull latest Node code from git master branch'
echo '#################################################'

git fetch && git checkout master && git pull

echo '#################################################'
echo 'Fix keys dir permissions as needed'
echo '#################################################'

sudo chown -R root keys && sudo chmod 777 keys && sudo chmod 777 keys/backups

echo '#################################################'
echo 'Fix .data dir permissions as needed'
echo '#################################################'

sudo mkdir -p .data/rocksdb && sudo chown root .data/rocksdb && sudo chmod 777 .data/rocksdb

echo '#################################################'
echo 'Generate new self-signed TLS cert as needed for 127.0.0.1'
echo '#################################################'
rm -rf cert.*
[ ! -f ./cert.key ] && ./certgen.sh 127.0.0.1

echo '#################################################'
echo 'Start ntpd service container if needed'
echo '#################################################'

make ntpd-start

echo '#################################################'
echo 'Start upgraded Node'
echo '#################################################'

export COMPOSE_IGNORE_ORPHANS=true
docker-compose up -d


scripts/setup.sh
#!/usr/bin/env bash
set -e

# Copyright 2017-2018 Tierion
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#     http://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Tested with Google and Digital Ocean Ubuntu 16.04 LTS Virtual Machines
#
# You can run this startup script manually, by copying it to the host,
# or by issuing this curl command. Since this command pipes the script
# directly into a bash shell you should examine the script before running.
#
#   curl -sSL https://chainpoint-node.storage.googleapis.com/setup.sh | bash
#
# Digital Ocean provides good documentation on how to manually install
# Docker on their platform.
#
#   https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-on-ubuntu-16-04
#
# Pre-requisites:
# - 64-bit Ubuntu 16.04 server
# - Non-root user with sudo privileges

# Don't run this script more than once!
if [ -f /.chainpoint-installer-run ]; then
    echo "Looks like this script has already been run. Exiting!"
    exit 0
fi

# Make sure we're running on Ubuntu 16.04 (Xenial)
if [ "$(. /etc/os-release; echo $NAME)" != "Ubuntu" ]; then
  echo "Looks like you are not running this on an Ubuntu OS. Exiting!"
  exit 1
fi

if [ "$(. /etc/os-release; echo $UBUNTU_CODENAME)" != "xenial" ]; then
  echo "Looks like you are not running this on Ubuntu version 16.04 (Xenial). Exiting!"
  exit 1
fi

echo '#################################################'
echo 'Installing Docker'
echo '#################################################'
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
sudo apt-get update
apt-cache policy docker-ce
sudo apt-get install -y docker-ce make

echo '#################################################'
echo 'Allow current user to use Docker without "sudo"'
echo 'REQUIRES SSH session logout + login'
echo '#################################################'
sudo usermod -aG docker ${USER}

echo '#################################################'
echo 'Installing Docker Compose'
echo '#################################################'
sudo mkdir -p /usr/local/bin
sudo curl -s -L "https://github.com/docker/compose/releases/download/1.21.0/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
sudo chmod +x /usr/local/bin/docker-compose

echo '#################################################'
echo 'Downloading chainpoint-node Github Repository'
echo '#################################################'
if [ ! -d "~/chainpoint-node" ]; then
  cd ~ && git clone -b master https://github.com/chainpoint/chainpoint-node
fi

echo '#################################################'
echo 'Creating .env config file from .env.sample'
echo '#################################################'
cd ~/chainpoint-node && make build-config

echo '#################################################'
echo 'Creating swap file as needed'
echo '#################################################'

if free | awk '/^Swap:/ {exit !$2}'; then
    echo "An existing swap file was detected. Skipping..."
else
    echo "No swap file detected. Installing..."
    sudo fallocate -l 2G /swapfile
    sudo chmod 600 /swapfile
    sudo mkswap /swapfile
    sudo swapon /swapfile
    echo '/swapfile none swap sw 0 0' | sudo tee -a /etc/fstab
    sudo sysctl vm.swappiness=10
    echo 'vm.swappiness=10' | sudo tee -a /etc/sysctl.conf
    sudo sysctl vm.vfs_cache_pressure=50
    echo 'vm.vfs_cache_pressure=50' | sudo tee -a /etc/sysctl.conf
fi

echo '#################################################'
echo 'Docker and docker-compose installation completed!'
echo 'Please now exit and restart this SSH session'
echo 'before continuing with the README instructions.'
echo '#################################################'

sudo touch /.chainpoint-installer-run
